File: runtimes/node/src/auth/oauth-hosted.ts
// T056: OAuth 2.1 + PKCE hosted-redirect flow for the Node runtime (FR-014, FR-016).
// Used in streamable-http/hosted mode: the authorization uses a configured public redirect URI
// instead of a 127.0.0.1 loopback. Same PKCE machinery as the loopback flow.

import { randomBytes } from "node:crypto";

import type { OAuthConfig } from "../manifest.js";
import { buildAuthorizationUrl, generatePkce, type TokenSet } from "./oauth-loopback.js";

export interface HostedFlowState {
  authorizationUrl: string;
  verifier: string;
  state: string;
  redirectUri: string;
}

/** Begin the hosted flow: produce the URL to redirect the user to. The caller (hosted server)
 * completes it when the public redirect URI receives the code. */
export function beginHostedFlow(config: OAuthConfig): HostedFlowState {
  if (config.redirectMode !== "hosted" || !config.hostedRedirectUri) {
    throw new Error("hosted flow requires redirectMode 'hosted' and a hostedRedirectUri");
  }
  const { verifier, challenge } = generatePkce();
  const state = randomBytes(8).toString("hex");
  const authorizationUrl = buildAuthorizationUrl(config, config.hostedRedirectUri, challenge, state);
  return { authorizationUrl, verifier, state, redirectUri: config.hostedRedirectUri };
}

/** Complete the hosted flow given the code returned to the public redirect URI. */
export async function completeHostedFlow(
  config: OAuthConfig,
  flow: HostedFlowState,
  code: string,
  returnedState: string,
): Promise<TokenSet> {
  if (returnedState !== flow.state) throw new Error("state mismatch");
  const res = await fetch(config.tokenEndpoint, {
    method: "POST",
    headers: { "content-type": "application/x-www-form-urlencoded" },
    body: new URLSearchParams({
      grant_type: "authorization_code",
      code,
      redirect_uri: flow.redirectUri,
      client_id: config.clientId,
      code_verifier: flow.verifier,
    }).toString(),
  });
  const json = (await res.json()) as { access_token: string; refresh_token?: string; expires_in?: number };
  return {
    accessToken: json.access_token,
    refreshToken: json.refresh_token,
    expiresAt: Date.now() + (json.expires_in ?? 3600) * 1000,
  };
}

File: runtimes/java/src/main/java/wrapper/auth/OAuthHosted.java
package wrapper.auth;

import java.util.Map;

/**
 * T058: OAuth 2.1 + PKCE hosted-redirect flow for the Java runtime (FR-014, FR-016). Mirrors the
 * Node/Python hosted flow: authorization uses a configured public redirect URI.
 */
public final class OAuthHosted {
  private OAuthHosted() {}

  public static final class HostedFlowState {
    public final String authorizationUrl;
    public final String verifier;
    public final String state;
    public final String redirectUri;

    public HostedFlowState(String authorizationUrl, String verifier, String state, String redirectUri) {
      this.authorizationUrl = authorizationUrl;
      this.verifier = verifier;
      this.state = state;
      this.redirectUri = redirectUri;
    }
  }

  public static HostedFlowState beginHostedFlow(Map<String, Object> config) throws Exception {
    if (!"hosted".equals(config.get("redirectMode")) || config.get("hostedRedirectUri") == null) {
      throw new IllegalArgumentException(
          "hosted flow requires redirectMode 'hosted' and a hostedRedirectUri");
    }
    String[] pkce = OAuthLoopback.generatePkce();
    String state = Long.toHexString(new java.security.SecureRandom().nextLong());
    String redirectUri = String.valueOf(config.get("hostedRedirectUri"));
    String url = OAuthLoopback.buildAuthorizationUrl(config, redirectUri, pkce[1], state);
    return new HostedFlowState(url, pkce[0], state, redirectUri);
  }

  public static OAuthLoopback.TokenSet completeHostedFlow(
      Map<String, Object> config,
      HostedFlowState flow,
      String code,
      String returnedState,
      OAuthLoopback.TokenExchanger exchanger)
      throws Exception {
    if (!flow.state.equals(returnedState)) {
      throw new IllegalStateException("state mismatch");
    }
    return exchanger.exchange(code, flow.verifier, flow.redirectUri);
  }
}
